IdentityCore/src/cache/mac/MSIDMacAppCredentialCacheItem.m

```diff
@@ -0,0 +1,194 @@
+    }
+    
+    return self;
+}
+
++ (MSIDMacAppCredentialCacheItem *)sharedInstance
+{
+    static MSIDMacAppCredentialCacheItem *instance = nil;
+    
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        instance = [[self alloc] initPrivate];
+    });
+    
+    return instance;
+}
+
+- (instancetype)initWithJSONDictionary:(NSDictionary *)json
+                                 error:(NSError * __autoreleasing *)error
+{
+    if (!(self = [self initPrivate]))
+    {
+        return nil;
+    }
+    
+    for (NSString *tokenKey in json)
+    {
+        NSDictionary *rtDict = [json objectForKey:tokenKey];
+        
+        if (rtDict)
+        {
+            MSIDCredentialCacheItem *appToken = [[MSIDCredentialCacheItem alloc] initWithJSONDictionary:rtDict error:error];
+            
+            if (appToken)
+            {
+                MSIDDefaultCredentialCacheKey *key = [[MSIDDefaultCredentialCacheKey alloc] initWithHomeAccountId:appToken.homeAccountId                      environment:appToken.environment clientId:appToken.clientId credentialType:appToken.credentialType];
+                
+                key.familyId = appToken.familyId;
+                key.realm = appToken.realm;
+                key.target = appToken.target;
+                key.enrollmentId = appToken.enrollmentId;
+                
+                [self setAppCredential:appToken forKey:key];
+            }
+        }
+    }
+    
+    return self;
+}
+
+- (NSDictionary *)jsonDictionary
+{
+    NSDictionary *cacheObjects = [self getCopy];
+    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
+    
+    for (MSIDDefaultCredentialCacheKey *appTokenKey in cacheObjects)
+    {
+        NSString *key = [NSString stringWithFormat:@"%@%@%@", appTokenKey.account, keyDelimiter, appTokenKey.service];
+        MSIDCredentialCacheItem *appToken = [cacheObjects objectForKey:appTokenKey];
+        NSDictionary *atDict = [appToken jsonDictionary];
+        
+        if (atDict)
+        {
+            [dictionary setObject:atDict forKey:key];
+        }
+    }
+    
+    return dictionary;
+}
+
+- (void)setAppCredential:(MSIDCredentialCacheItem *)token forKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects setObject:token forKey:key];
+    });
+}
+
+-(MSIDCredentialCacheItem *)appCredentialForKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    __block MSIDCredentialCacheItem *appCredential = nil;
+    
+    dispatch_sync(self.queue, ^{
+        appCredential = [self.cacheObjects objectForKey:key];
+    });
+    
+    return appCredential;
+}
+
+- (void)removeAppCredentialForKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects removeObjectForKey:key];
+    });
+}
+
+- (void)mergeAppCredential:(MSIDMacAppCredentialCacheItem *)credential
+{
+    NSDictionary *copy = [credential getCopy];
```

Comment:
nit: seems a bit weird that copying credential returns a dictionary. Consider changing naming.